### **Project 1: Displaying Top Rated 25 Movies from IMDB**

In [1]:
import requests
from bs4 import BeautifulSoup
import pandas as pd 

In [2]:
url='https://www.imdb.com/chart/top/'
r=requests.get(url)
print(r)

<Response [403]>


In [3]:
url='https://www.imdb.com/chart/top/'
r=requests.get(url)
print(r)
soup=BeautifulSoup(r.text,'lxml')
print(soup)

<Response [403]>
<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
</body>
</html>



When you get a 403 Forbidden response, it typically means that the server is rejecting your request. This could be because the server detects that you're making an automated request or that your headers do not adequately mimic a real browser.

**Methods to Bypass a 403 Error**   

Here are several methods you can try to bypass the **403 Forbidden** response and successfully scrape data using **eautifulSoup**:

**Use a Realistic User-Agent and Headers**


In [6]:
import requests
from bs4 import BeautifulSoup
import pandas
Movie_Name=[]
Year=[]
duration=[]
Rated=[]
Ratings=[]
Rating_Votecount=[]


url='https://www.imdb.com/chart/top/'
headers={'user-agent':
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko)Chrome/127.0.0.0 Safari/537.36',
    'accept-language':'en-US,en;q=0.9',
    'referer':'https://www.imdb.com/chart/top/',
    'Connection': 'keep-alive'        
            }
r=requests.get(url,headers=headers)
print(f"Status Code: {r.status_code}")   

if r.status_code==200:
        soup=BeautifulSoup(r.text,'lxml')
        Name=soup.find_all('h3',class_='ipc-title__text')[1:26]
        for i in Name:
            n=i.text
            Movie_Name.append(n)

        Year_elements = soup.find_all('div', class_='sc-b189961a-7 btCcOY cli-title-metadata')[:25]  # Replace with your actual selector
           # Iterate through each item and extract the last sub-element
        for item in Year_elements:
            sub_elements = item.find_all('span', class_='sc-b189961a-8 hCbzGp cli-title-metadata-item')[:25]
            if sub_elements:
                First_element = sub_elements[0].text.strip()
                Year.append(First_element)     

           # Assuming `items` is a list of elements where each element has multiple sub-elements
        durations = soup.find_all('div', class_='sc-b189961a-7 btCcOY cli-title-metadata')[:25]  # Replace with your actual selector
        
        # Iterate through each item and extract the last sub-element
        for item in durations:
            sub_elements = item.find_all('span', class_='sc-b189961a-8 hCbzGp cli-title-metadata-item')[:25]
            if sub_elements:
                Middle_element = sub_elements[1].text.strip()
                duration.append(Middle_element)

        # Assuming `items` is a list of elements where each element has multiple sub-elements
        items = soup.find_all('div', class_='sc-b189961a-7 btCcOY cli-title-metadata')[:25]  # Replace with your actual selector
        
        # Iterate through each item and extract the last sub-element
        for item in items:
            sub_elements = item.find_all('span', class_='sc-b189961a-8 hCbzGp cli-title-metadata-item')[:25]
            if sub_elements:
                last_element = sub_elements[-1].text.strip()
                Rated.append(last_element)
               

        Rating=soup.find_all('span',class_='ipc-rating-star--rating')[:25]
        for i in Rating:
            p=i.text
            Ratings.append(p)
        

        vote_count_elements = soup.find_all('span', class_='ipc-rating-star--voteCount')[:25]
        
        for i in vote_count_elements:
            text = i.text.replace('(',"").replace(')','').strip()
            Rating_Votecount.append(text)    


data = {
    'Movie_Name': Movie_Name,
    'Year': Year,
    'Duration': duration,
    'Rated': Rated,
    'Ratings': Ratings,
    'Rating_Votecount': Rating_Votecount
}

# Convert the dictionary into a DataFrame
df = pd.DataFrame(data)

# Display the DataFrame
print(df)

Status Code: 200
                                           Movie_Name  Year Duration  \
0                         1. The Shawshank Redemption  1994   2h 22m   
1                                    2. The Godfather  1972   2h 55m   
2                                  3. The Dark Knight  2008   2h 32m   
3                            4. The Godfather Part II  1974   3h 22m   
4                                     5. 12 Angry Men  1957   1h 36m   
5                                 6. Schindler's List  1993   3h 15m   
6    7. The Lord of the Rings: The Return of the King  2003   3h 21m   
7                                     8. Pulp Fiction  1994   2h 34m   
8   9. The Lord of the Rings: The Fellowship of th...  2001   2h 58m   
9                  10. The Good, the Bad and the Ugly  1966   2h 58m   
10                                   11. Forrest Gump  1994   2h 22m   
11          12. The Lord of the Rings: The Two Towers  2002   2h 59m   
12                                     13. Figh

In [8]:
df.to_csv("Top Rated 25 Movies from IMDB.csv")